specification/flag-evaluation/hooks.md

```diff
@@ -2,4 +2,126 @@
+
+> Hook context **SHOULD** provide: provider, client
+
+> flag key, flag type, default value properties **MUST** be immutable.
+
+> The evaluation context **MUST** be mutable only within the `before` hook.
+
+### State
+
+> State **MUST** be a map of objects.
+
+> Condition: State **MUST** be immutable, if the implementation language supports a mechanism for marking data as immutable.
+
+### Hook creation and parameters
+
+> Hooks **MUST** specify at least one stage.
+
+> The `before` stage *MUST* run before flag evaluation occurs. It accepts a `hook context` (required) and `state` (optional) as parameters and returns either a `HookContext` or nothing.
+
+```
+HookContext|void before(HookContext, State)
+```
+
+> The `after` stage **MUST** run after flag evaluation occurs. It accepts a `hook context` (required), `flag evaluation details` (required) and `state` (optional). It has no return value.
+
+> The `error` hook runs when errors are encountered in the `before` or `after` state or when flag evaluation errors. It accepts `hook context` (required), `exception` for what went wrong (required), and `state` (optional). It has no return value.
+
+> The `finally` hook **MUST** run after the `before`, `after`, and `error` stages. It accepts a `hook context` (required) and `state` (optional). There is no return value.
+
+> Condition: If `finally` is a reserved word in the language, `afterAll` should be used.
+
+### Hook registration & ordering
+
+> The API, Client and invocation **MUST** have a method for registering hooks which accepts `flag evaluation options`
+
+```js
+OpenFeature.addHooks(new Hook1());
+
+//...
+
+Client client = OpenFeature.getClient();
+client.addHooks(new Hook2());
+
+//...
+
+client.getValue('my-flag', 'defaultValue', new Hook3());
+```
+
+> Hooks **MUST** be evaluated in the following order: API before, client before, invocation before, invocation after, client after, API after, invocation finally, client finally, API finally. If errors are present, the order is: invocation error, client error, API error.
+
+```
+// here we are using the order defined above (note if it wasn't defined, we'd have to order them somehow anyway
+Hooks[] allHooks = [ ...OpenFeature.getHooks(), ...this.getHooks(), ...evaluationOptions.hooks ]
+
+try {
+  // evaluation all our before hooks
+  for (h in allHooks) {
+    h.before(...);
+  }
+
+  provider.resolveValue(...);
+
+  // evaluation all our after hooks
+  for (h in allHooks.reverse()) {
+    h.after(...);
+  }
+} catch (Error err) {
+  // evaluation all our error hooks
+  for (h in allHooks.reverse()) {
+    h.error(...);
+  }
+} finally {
+  // evaluation all our finally hooks
+  for (h in allHooks.reverse()) {
+    h.finally(...);
+  }
+}
+```
+
+> When an error occurs in hook evaluation, further evaluation **MUST** stop and error hook evaluation begins.
+
+> If an error is encountered in the error stage, it **MUST** not be returned to the user.
```

Comment:
for consistency, it seems like this should also be the case for errors in the finally stage?

Reply:
The intent of this line was:  If you have an error from the provider and that somehow triggers an error in the error flow.. don't overwrite their original error with this new one.

I'm not actually clear what an error in finally should do. Our whole MO seems to be "never throw, except maybe around client initialization", but also I don't want to have this logic complexity in the finally handler.

```python
try:
  before()
  eval()
  after()
except:
  error()
finally:
  try:
    finally()
  except:
    error()
```